sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -368,7 +369,7 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
 
   override def visitDeleteFromTable(
       ctx: DeleteFromTableContext): LogicalPlan = withOrigin(ctx) {
-    val table = createUnresolvedRelation(ctx.multipartIdentifier())
+    val table = createUnresolvedRelation(ctx.relationReference())
```

Comment:
This won't work for commands in the legacy style, which does not keep the table/view/namespace as a child logical plan. `CreateTableLikeCommand` is one example, and there can be more in third-party Spark forks.

Reply:
OK, alternatives?

Reply:
The only other idea I can think of is my below alternative to leave the multipartIdentifier references where they are, and just update its definition instead:

```
multipartIdentifier
    : IDENTIFIER_KW LEFT_PAREN expression RIGHT_PAREN
    | parts+=errorCapturingIdentifier (DOT parts+=errorCapturingIdentifier)*
    ;
```

We don't have to do that though to address this particular point, if there is another way.

Reply:
Right now visitMultipartIdentifer returns a Seq[String]. This change would force a change. I fear that is the change Wenchen is worried about (?).
Also, note that there are still 30 multipartIdentifier usages remaining which do NOT support IDENTIFIER() notation.
So I think we would trade mechanical churn in the grammar for code changes in AstBuilder et al.

Reply:
We can add a new parser rule as this PR already did
```
identifierReference:
    : IDENTIFIER_KW LEFT_PAREN expression RIGHT_PAREN
    | multipartIdentifier
```

then we update the parser rules for commands one by one, for each command, we update the corresponding scala part. Taking DROP TABLE as an example, the current scala part is
```
  override def visitDropTable(ctx: DropTableContext): LogicalPlan = withOrigin(ctx) {
    // DROP TABLE works with either a table or a temporary view.
    DropTable(
      UnresolvedIdentifier(visitMultipartIdentifier(ctx.multipartIdentifier()), allowTemp = true),
      ctx.EXISTS != null,
      ctx.PURGE != null)
  }
```
we can change it to
```
case class PlanWithUnresolvedIdentifier(identifierExpr: Expression, planBuilder: Seq[String] => LogicalPlan)

def withIdentClause(ctx: IdentifierReferenceContext, builder: Seq[String] => LogicalPlan): LogicalPlan = {
  if (ctx.expression != null) {
    PlanWithUnresolvedIdentifier(expr(ctx.expression), builder)
  } else {
    builder.apply(visitMultipartIdentifier(ctx.multipartIdentifier()))
  }
}

override def visitDropTable(ctx: DropTableContext): LogicalPlan = withOrigin(ctx) {
  withIdentClause(ctx.identifierReference, ident => {
        DropTable(
          UnresolvedIdentifier(ident, allowTemp = true),
          ctx.EXISTS != null,
          ctx.PURGE != null)
  }) 
}
```

Reply:
@cloud-fan Do you want this for commands only, or should I try the same for column, function, table references in queries, INSERT, etc...

Reply:
This is for commands only, and we can do similar things for expressions (column, function, etc.)
```
case class ExprWithUnresolvedIdentifier(identifierExpr: Expression, exprBuilder: Seq[String] => Expression)
```